File: android/app/src/main/java/io/github/cidy02/kudos/ui/components/ScopedRemovalBulkActionBar.kt
package io.github.cidy02.kudos.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.outlined.Bookmark
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.CollectionsBookmark
import androidx.compose.material.icons.outlined.Queue
import androidx.compose.material.icons.outlined.RemoveCircleOutline
import androidx.compose.material.icons.outlined.Star
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import io.github.cidy02.kudos.core.model.SavedWork
import io.github.cidy02.kudos.core.model.WorkCollection
import io.github.cidy02.kudos.core.model.ReadingQueue
import io.github.cidy02.kudos.library.ReadingQueueRepository
import io.github.cidy02.kudos.works.WorkRepository
import kotlinx.coroutines.launch

/**
 * The bulk-action bar for a single Reading Queue's or Collection's selection mode.
 * Shaped like [WorkBulkActionBar] (an icon row of secondary actions plus an
 * overflow menu), but the primary action is scoped membership removal rather than
 * a library-wide soft-delete — removing works from one queue/collection doesn't
 * delete or unsave them, so it gets its own confirmation copy and calls back into
 * [onRemove] (the caller's own scoped-removal loop) instead of
 * `workRepository.softDelete`.
 */
@Composable
fun ScopedRemovalBulkActionBar(
    selectedWorks: List<SavedWork>,
    workRepository: WorkRepository,
    queueRepository: ReadingQueueRepository?,
    /** e.g. "Remove from Queue" / "Remove from Collection". */
    removeLabel: String,
    /** e.g. "queue" / "collection" — used in the confirmation copy. */
    scopeName: String,
    onRemove: suspend () -> Unit,
    onDone: () -> Unit = {},
    modifier: Modifier = Modifier
) {
    val scope = rememberCoroutineScope()
    var confirmRemove by remember { mutableStateOf(false) }
    var showingAddToQueue by remember { mutableStateOf(false) }
    var showingAddToCollection by remember { mutableStateOf(false) }
    var overflowExpanded by remember { mutableStateOf(false) }

    val hasSelection = selectedWorks.isNotEmpty()
    val allFavorited = hasSelection && selectedWorks.all { it.isFavorite }
    val allSaved = hasSelection && selectedWorks.all { it.isSaved }
    val allFinished = hasSelection && selectedWorks.all { it.isFinished }

    Surface(
        tonalElevation = 6.dp,
        shadowElevation = 8.dp,
        modifier = modifier.fillMaxWidth()
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 8.dp, vertical = 4.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                IconButton(onClick = { confirmRemove = true }, enabled = hasSelection) {
                    Icon(Icons.Outlined.RemoveCircleOutline, removeLabel)
                }
                IconButton(onClick = {
                    scope.launch {
                        val target = !allFavorited
                        selectedWorks.forEach { workRepository.setFavorite(it.id, target) }
                        onDone()
                    }
                }, enabled = hasSelection) {
                    Icon(Icons.Outlined.Star, "Favorite")
                }
                IconButton(onClick = {
                    scope.launch {
                        val target = !allSaved
                        selectedWorks.forEach { workRepository.setSaved(it.id, target) }
                        onDone()
                    }
                }, enabled = hasSelection) {
                    Icon(Icons.Outlined.Bookmark, "Save")
                }
                IconButton(onClick = { showingAddToQueue = true }, enabled = hasSelection) {
                    Icon(Icons.Outlined.Queue, "Add to Queue")
                }
                IconButton(onClick = { showingAddToCollection = true }, enabled = hasSelection) {
                    Icon(Icons.Outlined.CollectionsBookmark, "Add to Collection")
                }
                IconButton(onClick = {
                    scope.launch {
                        val target = !allFinished
                        selectedWorks.forEach { workRepository.setFinished(it.id, target) }
                        onDone()
                    }
                }, enabled = hasSelection) {
                    Icon(Icons.Outlined.CheckCircle, "Mark Finished")
                }
            }

            Row(verticalAlignment = Alignment.CenterVertically) {
                Box {
                    IconButton(onClick = { overflowExpanded = true }, enabled = hasSelection) {
                        Icon(Icons.Default.MoreVert, "More actions")
                    }
                    DropdownMenu(
                        expanded = overflowExpanded,
                        onDismissRequest = { overflowExpanded = false }
                    ) {
                        DropdownMenuItem(
                            text = { Text("Unfavorite") },
                            onClick = {
                                overflowExpanded = false
                                scope.launch {
                                    selectedWorks.forEach { workRepository.setFavorite(it.id, false) }
                                    onDone()
                                }
                            }
                        )
                        DropdownMenuItem(
                            text = { Text("Unsave") },
                            onClick = {
                                overflowExpanded = false
                                scope.launch {
                                    selectedWorks.forEach { workRepository.setSaved(it.id, false) }
                                    onDone()
                                }
                            }
                        )
                        DropdownMenuItem(
                            text = { Text("Mark Unfinished") },
                            onClick = {
                                overflowExpanded = false
                                scope.launch {
                                    selectedWorks.forEach { workRepository.setFinished(it.id, false) }
                                    onDone()
                                }
                            }
                        )
                        HorizontalDivider()
                        DropdownMenuItem(
                            text = { Text(removeLabel, color = MaterialTheme.colorScheme.error) },
                            onClick = {
                                overflowExpanded = false
                                confirmRemove = true
                            }
                        )
                    }
                }
                TextButton(onClick = onDone) {
                    Text("Cancel")
                }
            }
        }
    }

    DestructiveConfirmation(
        show = confirmRemove,
        title = if (selectedWorks.size == 1) "$removeLabel?" else "$removeLabel (${selectedWorks.size})?",
        text = "The selected works will no longer be in this $scopeName. They stay in your Library either way.",
        confirmText = "Remove",
        confirmBeforeDelete = true,
        onConfirm = {
            confirmRemove = false
            scope.launch {
                onRemove()
                onDone()
            }
        },
        onDismissRequest = { confirmRemove = false }
    )

    if (showingAddToQueue) {
        var queues by remember { mutableStateOf<List<ReadingQueue>?>(null) }
        LaunchedEffect(Unit) {
            queues = queueRepository?.listQueues() ?: emptyList()
        }
        AlertDialog(
            onDismissRequest = { showingAddToQueue = false },
            title = { Text("Add Selection to Queue") },
            text = {
                Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                    if (queues == null) {
                        Text("Loading...", style = MaterialTheme.typography.bodyMedium)
                    } else if (queues!!.isEmpty()) {
                        Text("No custom queues yet.", style = MaterialTheme.typography.bodyMedium)
                    } else {
                        queues!!.forEach { queue ->
                            TextButton(
                                onClick = {
                                    showingAddToQueue = false
                                    scope.launch {
                                        selectedWorks.forEach { runCatching { queueRepository?.addWork(queue.id, it.id) } }
                                        onDone()
                                    }
                                },
                                modifier = Modifier.fillMaxWidth()
                            ) {
                                Text(queue.displayName, modifier = Modifier.fillMaxWidth())
                            }
                        }
                    }
                }
            },
            confirmButton = { TextButton(onClick = { showingAddToQueue = false }) { Text("Close") } }
        )
    }

    if (showingAddToCollection) {
        var newName by remember { mutableStateOf("") }
        var collections by remember { mutableStateOf<List<WorkCollection>?>(null) }
        LaunchedEffect(Unit) {
            collections = workRepository.allCollections()
        }
        AlertDialog(
            onDismissRequest = { showingAddToCollection = false },
            title = { Text("Add Selection to Collection") },
            text = {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    OutlinedTextField(
                        value = newName,
                        onValueChange = { newName = it },
                        label = { Text("New or existing collection") },
                        singleLine = true,
                        modifier = Modifier.fillMaxWidth()
                    )
                    if (collections != null && collections!!.isNotEmpty()) {
                        Text("Existing collections:", style = MaterialTheme.typography.labelMedium)
                        collections!!.forEach { col ->
                            TextButton(
                                onClick = {
                                    showingAddToCollection = false
                                    scope.launch {
                                        selectedWorks.forEach { runCatching { workRepository.addToCollection(it.id, col.name) } }
                                        onDone()
                                    }
                                },
                                modifier = Modifier.fillMaxWidth()
                            ) {
                                Text(col.name, modifier = Modifier.fillMaxWidth())
                            }
                        }
                    }
                }
            },
            confirmButton = {
                TextButton(
                    enabled = newName.trim().isNotEmpty(),
                    onClick = {
                        val name = newName.trim()
                        showingAddToCollection = false
                        scope.launch {
                            selectedWorks.forEach { runCatching { workRepository.addToCollection(it.id, name) } }
                            onDone()
                        }
                    }
                ) { Text("Add") }
            },
            dismissButton = { TextButton(onClick = { showingAddToCollection = false }) { Text("Cancel") } }
        )
    }
}

File: kudos-ao3-reader/UIComponents/ScopedRemovalBulkActionBar.swift
import SwiftData
import SwiftUI

/// The bulk-action bar for a single Reading Queue's or Collection's selection mode.
/// Shaped like `WorkBulkActionBar` (an "Actions" menu of secondary actions plus a
/// Done checkmark), but the primary action is scoped membership removal rather than
/// a library-wide delete — removing works from one queue/collection doesn't delete
/// or unsave them, so it gets its own confirmation copy and its own removal call
/// rather than reusing `WorkBulkActionBar`'s hardcoded Delete/soft-delete path.
struct ScopedRemovalBulkActionBar: View {
    let selectedWorks: [SavedWork]
    /// e.g. "Remove from Queue" / "Remove from Collection".
    let removeLabel: String
    /// e.g. "queue" / "collection" — used in the confirmation copy.
    let scopeName: String
    /// Performs the scoped removal for every selected work. Called only after the
    /// user confirms.
    var onRemove: () -> Void
    /// Called after a confirmed removal, and when the checkmark exits selection
    /// mode without removing anything.
    var onDone: () -> Void = {}

    @Environment(\.modelContext) private var context
    @State private var confirmRemove = false
    @State private var showingAddToQueue = false
    @State private var showingAddToCollection = false
    @State private var showingTagSheet = false
    /// 1bg's Move to. Same destination picker as Add to Queue — the difference is
    /// what happens after, so it is one flag rather than a second sheet.
    @State private var showingMoveToQueue = false
    /// Queue-membership count across the selection, sampled when the move picker
    /// opens. The picker is shared with Add to Queue and cannot tell this bar
    /// whether anything was chosen, so a plain `onDismiss: onRemove` would strip
    /// the works out of this queue even when the reader pressed Cancel.
    @State private var membershipBeforeMove = 0
    @State private var isDownloading = false

    private var allSaved: Bool {
        !selectedWorks.isEmpty && selectedWorks.allSatisfy(\.isSaved)
    }

    private var allFavorited: Bool {
        !selectedWorks.isEmpty && selectedWorks.allSatisfy(\.isFavorite)
    }

    private var allSavedForLater: Bool {
        !selectedWorks.isEmpty && selectedWorks.allSatisfy(\.isInSavedForLaterQueue)
    }

    private var allFinished: Bool {
        !selectedWorks.isEmpty && selectedWorks.allSatisfy(\.isFinished)
    }

    var body: some View {
        Button(role: .destructive) {
            confirmRemove = true
        } label: {
            Label(removeLabel, systemImage: "minus.circle")
        }
        .disabled(selectedWorks.isEmpty)

        Spacer()

        Menu {
            Button {
                bulkSave()
            } label: {
                Label(
                    WorkActionLabels.saved(isSaved: allSaved).title,
                    systemImage: WorkActionLabels.saved(isSaved: allSaved).systemImage
                )
            }
            Button {
                bulkFavorite()
            } label: {
                Label(allFavorited ? "Favorited" : "Favorite", systemImage: allFavorited ? "star.fill" : "star")
            }
            Button {
                bulkToggleSavedForLater()
            } label: {
                Label(
                    WorkActionLabels.savedForLater(isQueued: allSavedForLater).title,
                    systemImage: WorkActionLabels.savedForLater(isQueued: allSavedForLater).systemImage
                )
            }
            Button {
                showingAddToQueue = true
            } label: {
                Label("Add to Queue", systemImage: "list.bullet.rectangle")
            }
            Button {
                showingAddToCollection = true
            } label: {
                Label("Add to Collection", systemImage: "square.stack")
            }
            // 1bg's Move to. Its own note calls this two verbs — add AND remove —
            // which is why it reuses the destination picker and then runs the
            // scoped removal this bar already owns, rather than being a variant
            // of Add to Queue that quietly leaves the work in both places.
            Button {
                membershipBeforeMove = totalQueueMemberships
                showingMoveToQueue = true
            } label: {
                Label("Move to Queue", systemImage: "arrow.right.square")
            }
            // 1bg names Tag as one of the four things a queue can do to a
            // selection. The sheet already existed for 1af and is already wired
            // into Library's own bar; both surfaces that mount this bar — a queue
            // and a collection — hold local works, so it applies to each.
            Button {
                showingTagSheet = true
            } label: {
                Label("Tag", systemImage: "tag")
            }
            // 1bg's Download. The "Download" item above it sets `isSaved`, which
            // only stops an EPUB being freed — for a work whose copy is already
            // gone it changes a flag and downloads nothing. This fetches.
            Button {
                Task { await bulkDownload() }
            } label: {
                Label("Download missing copies", systemImage: WorkActionLabels.downloadEmptySymbol)
            }
            .disabled(isDownloading || missingCopies.isEmpty)
            Button {
                bulkToggleFinished()
            } label: {
                let labels = WorkActionLabels.finished(isFinished: allFinished)
                Label(labels.title, systemImage: labels.systemImage)
            }
        } label: {
            Text("Actions")
                .font(.subheadline.weight(.medium))
                .padding(.horizontal, 14)
                .padding(.vertical, 6)
                .background(.regularMaterial, in: Capsule())
        }
        .disabled(selectedWorks.isEmpty)

        Spacer()

        Button {
            onDone()
        } label: {
            Image(systemName: "checkmark")
        }
        .accessibilityLabel("Done")
        .sheet(isPresented: $showingAddToQueue) {
            AddToQueueView(works: selectedWorks)
        }
        // The removal runs on dismiss rather than inside the picker: the picker
        // is shared with Add to Queue and must not learn about this scope. It
        // runs ONLY if the selection actually joined something, so cancelling
        // the picker leaves the works exactly where they were.
        .sheet(isPresented: $showingMoveToQueue, onDismiss: {
            guard totalQueueMemberships > membershipBeforeMove else { return }
            onRemove()
        }) {
            AddToQueueView(works: selectedWorks)
        }
        .sheet(isPresented: $showingAddToCollection) {
            AddToCollectionView(works: selectedWorks)
        }
        .sheet(isPresented: $showingTagSheet) {
            WorkBulkTagSheet(works: selectedWorks)
        }
        .confirmationDialog(
            "Remove \(selectedWorks.count) work\(selectedWorks.count == 1 ? "" : "s")?",
            isPresented: $confirmRemove,
            titleVisibility: .visible
        ) {
            Button("Remove", role: .destructive) {
                onRemove()
                onDone()
            }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("The selected works will no longer be in this \(scopeName). "
                + "They stay in your Library either way.")
        }
    }

    private var totalQueueMemberships: Int {
        selectedWorks.reduce(0) { total, work in
            total + work.queueMemberships.filter { !$0.isPendingDeletion }.count
        }
    }

    /// Works whose EPUB is gone — the only ones a download can do anything for.
    /// Re-fetching a work that already has one would reset its reading position
    /// for no gain.
    private var missingCopies: [SavedWork] {
        selectedWorks.filter { !WorkReaderPreparation.hasReadableEPUB(for: $0) }
    }

    /// Sequential, not concurrent: this is the same AO3 endpoint per work, and a
    /// parallel burst over a large selection is the kind of thing that gets an
    /// app rate-limited. A failure on one work does not stop the rest — the work
    /// simply still has no copy, which is the state it was already in.
    private func bulkDownload() async {
        isDownloading = true
        defer { isDownloading = false }
        for work in missingCopies {
            try? await WorkReaderPreparation.restoreReadableEPUB(for: work, in: context)
        }
    }

    private func bulkSave() {
        let shouldSave = !allSaved
        for work in selectedWorks {
            WorkLifecycle.setSaved(work, shouldSave, in: context)
        }
    }

    private func bulkFavorite() {
        let shouldFavorite = !allFavorited
        let now = Date()
        for work in selectedWorks {
            work.isFavorite = shouldFavorite
            work.markModified(now)
        }
        try? context.save()
    }

    private func bulkToggleSavedForLater() {
        let shouldSave = !allSavedForLater
        for work in selectedWorks {
            if shouldSave {
                guard !work.isInSavedForLaterQueue else { continue }
                Task { @MainActor in
                    _ = await ReadingQueueService.addToSavedForLater(work, in: context)
                }
            } else {
                guard work.isInSavedForLaterQueue else { continue }
                ReadingQueueService.removeFromQueueAndDeleteIfQueueOnly(
                    work,
                    from: ReadingQueueService.ensureSavedForLaterQueue(in: context),
                    in: context
                )
            }
        }
    }

    private func bulkToggleFinished() {
        let shouldFinish = !allFinished
        for work in selectedWorks {
            if shouldFinish {
                WorkLifecycle.markFinished(work, in: context)
            } else {
                WorkLifecycle.markStillReading(work, in: context)
            }
        }
    }
}
